File: government/src/main/java/rest/InvoiceResource.java
package rest;

import domain.Invoice;
import domain.enums.InvoiceGenerationType;
import domain.enums.InvoiceStatus;
import service.InvoiceService;

import javax.ejb.Stateless;
import javax.inject.Inject;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import java.util.Date;
import java.util.List;

@Path("invoices")
@Stateless
public class InvoiceResource {

    @Inject
    InvoiceService invoiceService;

    @GET
    @Path("{id}")
    @Produces("application/json")
    public Invoice getInvoiceById(@PathParam("id") Long id){

        return invoiceService.getInvoiceById(id);
    }

    @GET
    @Produces("application/json")
    public List<Invoice> getAllInvoices(){

        return invoiceService.getAllInvoices();
    }

    @GET
    @Path("/vehicle/{id}")
    @Produces("application/json")
    public List<Invoice> getAllInvoicesByVehicleId(@PathParam("id") Long id) {

        return invoiceService.getAllInvoicesByVehicle(id);
    }

    @GET
    @Path("/civilian/{id}")
    @Produces("application/json")
    public List<Invoice> getAllInvoicesByCivilianId(@PathParam("id") Long id) {

        return invoiceService.getAllInvoicesByCivilian(id);
    }

    @GET
    @Path("/date/created/{startdate}/{enddate}")
    @Produces("application/json")
    public List<Invoice> getAllInvoicesCreatedBetweenDates(@PathParam("startdate")Date startDate,
                                                           @PathParam("enddate")Date endDate){

        return invoiceService.getAllInvoicesCreatedBetweenDates(startDate, endDate);
    }

    @GET
    @Path("/date/for/{startdate}/{enddate}")
    @Produces("application/json")
    public List<Invoice> getAllInvoicesForBetweenDates(@PathParam("startdate")Date startDate,
                                                       @PathParam("enddate")Date endDate){

        return invoiceService.getAllInvoicesForBetweenDates(startDate, endDate);
    }

    @GET
    @Path("/generated/{type}")
    @Produces("application/json")
    public List<Invoice> getAllInvoicesGeneratedBy(@PathParam("type")InvoiceGenerationType type){

        return invoiceService.getAllInvoicesGeneratedBy(type);
    }

    @GET
    @Path("/status/{status}")
    @Produces("application/json")
    public List<Invoice> getAllInvoicesByStatus(@PathParam("status")InvoiceStatus status){

        return invoiceService.getAllInvoicesByStatus(status);
    }
}

File: src/main/java/rest/InvoiceResource.kt
package rest

import domain.Invoice
import domain.enums.InvoiceGenerationType
import domain.enums.InvoiceState
import service.InvoiceService
import utils.Open
import java.util.Date
import javax.inject.Inject
import javax.ws.rs.GET
import javax.ws.rs.POST
import javax.ws.rs.PUT
import javax.ws.rs.Path
import javax.ws.rs.PathParam
import javax.ws.rs.Produces
import javax.ws.rs.core.Context
import javax.ws.rs.core.Response
import javax.ws.rs.core.UriInfo

@Path("invoices")
@Open
class InvoiceResource @Inject constructor(
    private val invoiceService: InvoiceService
) {
    @Context
    private lateinit var request: UriInfo

    @GET
    @Produces("application/json")
    fun allInvoices(): List<Invoice> {
        val startDate: Long = request.queryParameters.get("startDate")?.first()?.toLongOrNull() ?: 0
        val endDate: Long = request.queryParameters.get("endDate")?.first()?.toLongOrNull() ?: Date(3000, 1, 1).toInstant().toEpochMilli()
        return invoiceService.allInvoicesCreatedBetweenDates(startDate, endDate)
    }

    @GET
    @Path("{uuid}")
    @Produces("application/json")
    fun getInvoiceById(@PathParam("uuid") uuid: String): Response =
        Response.ok(invoiceService.getInvoiceByUuid(uuid)).build()

    @GET
    @Path("/vehicle/{id}")
    @Produces("application/json")
    fun allInvoicesByVehicleId(@PathParam("id") id: String): List<Invoice> =
        invoiceService.allInvoicesByVehicle(id)

    @GET
    @Path("/civilian/{id}")
    @Produces("application/json")
    fun allInvoicesByCivilianId(@PathParam("id") id: String): List<Invoice> =
        invoiceService.allInvoicesByCivilian(id)

    @GET
    @Path("/date/for/{start}/{end}")
    @Produces("application/json")
    fun allInvoicesForBetweenDates(
        @PathParam("start") start: String,
        @PathParam("end") end: String
    ): List<Invoice> =
        invoiceService.allInvoicesForBetweenDates(start, end)

    @GET
    @Path("/generated/{type}")
    @Produces("application/json")
    fun allInvoicesGeneratedBy(@PathParam("type") type: String): List<Invoice> =
        invoiceService.allInvoicesGeneratedBy(InvoiceGenerationType.valueOf(type))

    @GET
    @Path("/state/{state}")
    @Produces("application/json")
    fun allInvoicesByState(@PathParam("state") state: String): List<Invoice> =
        invoiceService.allInvoicesByState(InvoiceState.valueOf(state))

    @POST
    @Path("/update/state/{uuid}")
    @Produces("application/json")
    fun updateInvoiceState(@PathParam("uuid") uuid: String): Response {
        val invoice = invoiceService.updateInvoiceState(uuid, InvoiceState.valueOf(request.queryParameters.getFirst("state")))
            ?: return Response.notModified().build()

        return Response.ok(invoice).build()
    }

    @PUT
    @Path("/regenerate/{uuid}")
    @Produces("application/json")
    fun regenerateInvoice(@PathParam("uuid") uuid: String): Response {
        val regeneratedInvoice = invoiceService.regenerateInvoice(uuid)
            ?: return Response.notModified().build()

        return Response.ok(regeneratedInvoice).build()
    }
}
